ios/RNAudioRecorderPlayer.swift

```diff
@@ -159,166 +207,101 @@ class RNAudioRecorderPlayer: RCTEventEmitter, AVAudioRecorderDelegate {
+            AVLinearPCMBitDepthKey: audioSets["AVLinearPCMBitDepthKeyIOS"] as? Int ?? AVLinearPCMBitDepthKey.count,
+            AVLinearPCMIsBigEndianKey: audioSets["AVLinearPCMIsBigEndianKeyIOS"] as? Bool ?? true,
+            AVLinearPCMIsFloatKey: audioSets["AVLinearPCMIsFloatKeyIOS"] as? Bool ?? false,
+            AVLinearPCMIsNonInterleaved: audioSets["AVLinearPCMIsNonInterleavedIOS"] as? Bool ?? false,
+            AVEncoderBitRateKey: audioSets["AVEncoderBitRateKeyIOS"] as? Int ?? 128000
+        ] as [String: Any]
+
+        updateAudioFileURL(path: path, format: avFormat)
+        let avMode = avMode(fromString: audioSets["AVModeIOS"] as? String ?? "default") ?? .default
+        do {
+            try audioSession.setCategory(.playAndRecord, mode: avMode, options: [.defaultToSpeaker, .allowBluetooth])
+            try audioSession.setActive(true)
+        } catch {
+            completion(.failure(.audioSessionError(error)))
         }
-
-        if (mode == "measurement") {
-            avMode = AVAudioSession.Mode.measurement
-        } else if (mode == "gamechat") {
-            avMode = AVAudioSession.Mode.gameChat
-        } else if (mode == "movieplayback") {
-            avMode = AVAudioSession.Mode.moviePlayback
-        } else if (mode == "spokenaudio") {
-            avMode = AVAudioSession.Mode.spokenAudio
-        } else if (mode == "videochat") {
-            avMode = AVAudioSession.Mode.videoChat
-        } else if (mode == "videorecording") {
-            avMode = AVAudioSession.Mode.videoRecording
-        } else if (mode == "voicechat") {
-            avMode = AVAudioSession.Mode.voiceChat
-        } else if (mode == "voiceprompt") {
-            if #available(iOS 12.0, *) {
-                avMode = AVAudioSession.Mode.voicePrompt
-            } else {
-                // Fallback on earlier versions
+        audioSession.requestRecordPermission { granted in
+            DispatchQueue.main.async {
+                guard granted else { return completion(.failure(.recordingPermissionNotGranted)) }
+                guard let audioFileURL = self.audioFileURL else { return completion(.failure(.failedToLocateRecordingFile)) }
+                guard let audioRecorder = try? AVAudioRecorder(url: audioFileURL, settings: settings) else { return completion(.failure(.failedToCreateRecorder)) }
+
+                audioRecorder.prepareToRecord()
+                audioRecorder.delegate = self
+                audioRecorder.isMeteringEnabled = meteringEnabled
+                guard audioRecorder.record() else { return completion(.failure(.failedToStartRecording)) }
+
+                self.currentAudioRecorder = audioRecorder
+                self.recordingDidStart()
+                self.startRecorderTimer()
+                completion(.success(audioFileURL))
             }
         }
+    }
 
+    private func pauseCurrentRecording() throws {
```

Comment:
Feels a bit funny that e.g. `startNewRecording` is essentially async (has `completion` callback) while pause and resume aren't – can we unify maybe?